homeassistant/components/fan/zha.py

```diff
@@ -76,34 +74,36 @@ def is_on(self) -> bool:
             return False
         return self._state != SPEED_OFF
 
-    @asyncio.coroutine
-    def async_turn_on(self, speed: str = None, **kwargs) -> None:
+    async def async_turn_on(self, speed: str = None, **kwargs) -> None:
         """Turn the entity on."""
         if speed is None:
             speed = SPEED_MEDIUM
 
-        yield from self.async_set_speed(speed)
+        await self.async_set_speed(speed)
 
-    @asyncio.coroutine
-    def async_turn_off(self, **kwargs) -> None:
+    async def async_turn_off(self, **kwargs) -> None:
         """Turn the entity off."""
-        yield from self.async_set_speed(SPEED_OFF)
+        await self.async_set_speed(SPEED_OFF)
 
-    @asyncio.coroutine
-    def async_set_speed(self, speed: str) -> None:
+    async def async_set_speed(self, speed: str) -> None:
         """Set the speed of the fan."""
-        yield from self._endpoint.fan.write_attributes({
-            'fan_mode': SPEED_TO_VALUE[speed]})
+        from zigpy.exceptions import DeliveryError
+        try:
+            await self._endpoint.fan.write_attributes(
+                {'fan_mode': SPEED_TO_VALUE[speed]}
+            )
+        except DeliveryError as ex:
+            _LOGGER.error("%s: Could not set speed: %s", self.entity_id, ex)
+            return
 
         self._state = speed
         self.async_schedule_update_ha_state()
 
-    @asyncio.coroutine
-    def async_update(self):
+    async def async_update(self):
         """Retrieve latest state."""
-        result = yield from zha.safe_read(self._endpoint.fan, ['fan_mode'])
+        result = await zha.safe_read(self._endpoint.fan, ['fan_mode'])
         new_value = result.get('fan_mode', None)
-        self._state = VALUE_TO_SPEED.get(new_value, None)
+        self._state = VALUE_TO_SPEED.get(new_value, self._state)
```

Comment:
The state is defined on core and need to be in that format. Please revert this change.

Reply:
you lost me here for a bit. `self._state` is defined by ZhaFan and pushed as a state. This code just keeps the previous state in case we fail to read the `'fan_mode'` attribute. Although, from the other hand if we fail to get 'fan_mode' then the state should be None to represent the 'unknown'.
So is it better to keep old state or return None for failed updates?